src/java.base/share/classes/java/lang/ScopedValue.java

```diff
@@ -727,57 +738,99 @@ private static final class Cache {
+        //
+        // In more detail:
+        //
+        //  The size of the cache depends on System.getProperty. Generating the
+        //  hash of an instance of ScopedValue depends on ThreadLocalRandom.
+        //
+        //  Invoking either of these early in the JDK boot process will cause
+        //  startup to fail with an unrecoverable circular dependency.
+        //
+        // To break these cycles we allow scoped values to be created (but not
+        // used) without invoking either System.getProperty or
+        // ThreadLocalRandom. To do this we defer querying System.getProperty
+        // until the first reference to CACHE_TABLE_SIZE, and we define a local
+        // hashGenerator which is used until CACHE_TABLE_SIZE is initialized.
+
+        private static class Constants {
+            // The number of elements in the cache array, and a bit mask used to
+            // select elements from it.
+            private static final int CACHE_TABLE_SIZE, SLOT_MASK;
+            // The largest cache we allow. Must be a power of 2 and greater than
+            // or equal to 2.
+            private static final int MAX_CACHE_SIZE = 16;
+
+            private static final JavaUtilConcurrentTLRAccess THREAD_LOCAL_RANDOM_ACCESS
+                = SharedSecrets.getJavaUtilConcurrentTLRAccess();
+
+            static {
+                final String propertyName = "java.lang.ScopedValue.cacheSize";
+                var sizeString = System.getProperty(propertyName, "16");
+                var cacheSize = Integer.valueOf(sizeString);
+                if (cacheSize < 2 || cacheSize > MAX_CACHE_SIZE) {
+                    cacheSize = MAX_CACHE_SIZE;
+                    System.err.println(propertyName + " is out of range: is " + sizeString);
+                }
+                if ((cacheSize & (cacheSize - 1)) != 0) {  // a power of 2
+                    cacheSize = MAX_CACHE_SIZE;
+                    System.err.println(propertyName + " must be an integer power of 2: is " + sizeString);
+                }
+                CACHE_TABLE_SIZE = cacheSize;
+                SLOT_MASK = cacheSize - 1;
+
+                // hashGenerator is set here in order not to initialize
+                // j.u.c.ThreadLocalRandom early in the JDK boot
+                // process. After this static initialization block,
+                // new instances of ScopedValue will be initialized by a
+                // thread-local random generator.
+                hashGenerator = new IntSupplier() {
+                    @Override
+                    public int getAsInt() {
+                        int x;
+                        do {
+                            x = THREAD_LOCAL_RANDOM_ACCESS
+                                .nextSecondaryThreadLocalRandomSeed();
+                        } while (Cache.primarySlot(x) == Cache.secondarySlot(x));
+                        return x;
+                    }
+                };
```

Comment:
There's something a bit uncomfortable about initializing hashGenerator in ScopedValue's class initializer, then changing it in Constants class initializer. Iimmediately clear what the memory model issues. It doesn't of course matter if the stale value is used but I think confusing on first sight.

Have you tried dropping the initializer from ScopedValue so it hashGenerator is only set by Constants? The ScopedValue constructor can call generateKey if hashGenerator is null, and hashGenerator can become a stable field.

Reply:
> > Have you tried dropping the initializer from ScopedValue so hashGenerator is only set by Constants? The ScopedValue constructor can call generateKey if hashGenerator is null, and hashGenerator can become a stable field.
> 
> That would be more explicit.

A stable field would also be constant foldable.